### DISCUSSION TITLE
How to deploy Turbo repo with AWS Sam CLI?

### DISCUSSION THREAD

**Discussion by fnLog0** [aws/aws-sam-cli#6690] [Questions]
I have a TurboRepo for an AWS AppSync application, and I want to deploy it using AWS SAM. However, when I run sam build, it's using local packages from npm instead of taking them from the workspace. Can you please assist me with this?

https://github.com/vercel/turbo/discussions/7345

**Comment by uldyssian-sh** [comment]
SAM’s Node.js build is not workspace-aware by default.

When you run `sam build` for a Node.js Lambda, SAM uses its own build workflow (NodejsNpmBuilder / esbuild) and expects a normal “single package” layout. In a monorepo (TurboRepo, npm workspaces, pnpm, yarn workspaces), dependencies may be symlinked or hoisted and SAM’s builder can end up:
- resolving dependencies incorrectly
- ignoring workspace packages
- or packaging only what exists inside the function’s CodeUri

That’s why you see it “using local packages” incorrectly instead of the workspace graph.

The reliable way to deploy a TurboRepo with SAM is:

1) Build your workspace yourself (Turbo) before SAM packaging
2) Point SAM at the built artifacts (a standalone folder) rather than the raw workspace source

Recommended approach (works consistently):

Step 1: create a “deploy” build output per Lambda package

Use Turbo to build each Lambda into a dist folder, e.g.

turbo run build --filter=my-lambda

and make sure the build output includes either:
- bundled JS (single file) OR
- a pruned node_modules that contains all dependencies

Step 2: make SAM package only that output

In template.yaml set:

MyFunction:
  Type: AWS::Serverless::Function
  Properties:
    Runtime: nodejs20.x
    Handler: index.handler
    CodeUri: apps/my-lambda/dist/

Now SAM is not trying to understand your monorepo. It just uploads the already-built artifact.

Best practice: bundle dependencies

For Node.js, the cleanest monorepo story is to bundle (esbuild/webpack) so the function package is self-contained and does not rely on workspace symlinks.

Typical pattern:
- build step produces dist/index.js
- externalize only aws-sdk (or nothing, depending on runtime)
- CodeUri points to dist/
- Handler points to dist entry

Alternative approach: use “sam build --use-container” + single-package CodeUri

This still often breaks with workspaces because the builder stages CodeUri into a temp directory and loses the workspace symlinks.

So it may work for plain npm installs, but it is not a dependable solution for TurboRepo packages.

Summary:

SAM build is not monorepo/workspace aware.

Don’t rely on SAM to resolve workspace packages.

Instead, run Turbo to build a deployable artifact per Lambda (bundle or prune deps), then point SAM’s CodeUri at that build output and deploy.

**Reply by uldyssian-sh** [reply]
Thanks! If this helps, please consider marking it as “Helpful” or “Accepted” 😊
